packages/@glimmer/application/src/application.ts

```diff
@@ -231,38 +286,12 @@ export default class Application implements Owner {
    * at the end of the browser's event loop.
    */
   protected _rerender() {
-    let { env, _result: result } = this;
+    let { env  } = this;
 
-    if (!result) {
-      throw new Error('Cannot re-render before initial render has completed');
-    }
-
-    env.begin();
-    result.rerender();
-    env.commit();
-
-    this._didRender();
-  }
-
-  /** @hidden */
-  protected _render(): void {
-    let { env, templateIterator } = this;
-
-    // Begin a new transaction. The transaction stores things like component
-    // lifecycle events so they can be flushed once rendering has completed.
     env.begin();
-
-    // Iterate the template iterator, executing the compiled template program
-    // until there are no more instructions left to execute.
-    let result;
-    do {
-      result = templateIterator.next();
-    } while (!result.done);
-
-    // Finally, commit the transaction and flush component lifecycle hooks.
+    this.renderer.rerender();
```

Comment:
Re-render is synchronous but render is async? Seems like these should be balanced or at least check for the balancing.

Reply:
In practice, rerender cannot be async at the moment. The interface for `rerender()` already supports a Promise return type, but there aren't any implementations that can test this path yet.